linera-service/src/proxy/grpc.rs

```diff
@@ -468,6 +471,38 @@ where
         Ok(Response::new(certificate.try_into()?))
     }
 
+    #[instrument(skip_all, err(Display))]
+    async fn download_certificates(
+        &self,
+        request: Request<CertificatesBatchRequest>,
+    ) -> Result<Response<CertificatesBatchResponse>, Status> {
+        let hashes = request
+            .into_inner()
+            .hashes
+            .into_iter()
+            .map(linera_base::crypto::CryptoHash::try_from)
+            .collect::<Result<Vec<linera_base::crypto::CryptoHash>, _>>()?;
+
+        // Use 70% of the max message size as a buffer capacity.
+        // Leave 30% as overhead.
+        let mut grpc_message_limit: GrpcMessageLimiter<linera_chain::data_types::Certificate> =
+            GrpcMessageLimiter::new((GRPC_MAX_MESSAGE_SIZE as f64 * 0.7) as usize);
+
+        let certificates: Vec<linera_chain::data_types::Certificate> = self
+            .0
+            .storage
+            .read_certificates(hashes)
+            .await
+            .map_err(|err| Status::from_error(Box::new(err)))?
+            .into_iter()
+            .take_while(|cert| grpc_message_limit.fits::<Certificate>(cert.clone()))
```

Comment:
I have to clone a certificate since protobuf encoding needs to own the value being serialized.